### ISSUE TITLE
New Rule: disallow bidirectional control characters and homoglyphs to mitigate trojan source attacks

### ISSUE THREAD

**Issue by joyeecheung** [eslint/eslint#15240]
### Rule details

disallow bidirectional control characters and homoglyphs in source code to mitigate CVE-2021-42574 and CVE-2021-42694

### Related ECMAScript feature

Not a new feature, but to specifically disable certain characters supported by https://tc39.es/ecma262/#sec-unicode-format-control-characters

### What type of rule is this?

Warns about a potential problem

### Example code

```js
// From the trojan source paper: https://www.trojansource.codes/trojan-source.pdf

var accessLevel = "user";
if (accessLevel != "user{U+202E} {U+2066}// Check if admin{U+2069} {U+2066}") {
  console.log("You are an admin");
}

// Replacing {U+NNNN} in the code above with actual code points NNNN,
// the code would be rendered in bidirectional-aware tools as

if (accessLevel != "user") { // Check if admin
  console.log("You are an admin");
}

// Another variant (CVE-2021-42694) is to use homoglyphs
// (characters that look similar) to, for example, call into a
// different function
function sayHello() {
  console.log("Hello, World!");
}

// Replacing H(U+0048) above with Н(U+041D) would call into a
// different function, even though they look visually similar.
function sayНello() {
  console.log("Hello, World!");
}
```

### Why should this rule be in the core instead of a plugin?

These vuneralbilities have a relatively wide security impact affecting multiple programming languages, adding it in core would help promote awareness of these issues and reduce the risks in the JS ecosystem while the JS engines work out a fix or other alternatives to mitigate this.

### Participation

- [X] I am willing to submit a pull request to implement this rule.

### Additional comments

More information about these attacks can be found at https://www.trojansource.codes/. Also see the security announcement by Node.js: https://groups.google.com/g/nodejs-sec/c/_w6hoamG14E

Blog by the rust team about CVE-2021-42574: https://blog.rust-lang.org/2021/11/01/cve-2021-42574.html

**Comment by gabegorelick** [comment]
Some previous discussion on this: https://github.com/eslint/eslint/issues/10747

**Comment by nzakas** [comment]
We typically don’t add new rules for anything other than new language features, but considering this is a new vulnerability disclosed on 2021-11-01, I think this deserves consideration.

**Comment by mhofman** [comment]
I'm not sure the homoglyph or zero-width-joiner vector deserves a new rule. The `no-shadow` rule already exists and has the same spirit. It could be extended to match identifiers normalizing for these unicode characters. Potentially the `no-redeclare` rule if we want to keep `no-shadow`'s role to nested scopes.

**Comment by ljharb** [comment]
It's not just about identifiers tho, is it? These characters can be present anywhere, and `no-shadow` only deals in variables.

**Comment by mhofman** [comment]
My understanding of the paper is that the homoglyphs vector is only a risk for identifiers. The BiDi characters is another story, and I believe those would require parsing comments, which is out of scope IMO.

**Comment by ljharb** [comment]
It seems a lot less clear to me to bundle a security rule underneath a clarity rule.

**Comment by mhofman** [comment]
Fair, but the homoglyph vector is only an attack because it's not clear which identifier is referenced.

**Comment by joyeecheung** [comment]
The examples only demonstrate how these characters can be used for malicious intent. What the linter can do can be much simpler than trying to identify these attacks - just warns about the existence of these characters (which is what the GitHub UI and the rust linter currently do). With the checks applied to a project, the user can still choose to disable it occasionally when the characters are actually intended (e.g. having comments in RTL languages).

**Comment by mhofman** [comment]
Just to be clear, there are 2 vectors in the Trojan Source. One is BiDi characters, which GitHub warns about. The other is homoglyph between different character sets. There is no reason to disallow broadly non-latin character sets, only when used maliciously to confuse which identifier is referenced.

**Comment by joyeecheung** [comment]
Yes, the BiDi control characters is a very specific small set. Homoglyphs can be somewhat tricky to determine, though perhaps the Unicode confusables database can be a good reference. 

Regarding homoglyphs: I think it’s not just identifiers that can be affected. For example in the accessLevel example in the OP, replacing the “user” string content with some homoglyphs would be equally problematic. There might be more creative uses of this too.

**Comment by mdjermanovic** [comment]
> Homoglyphs can be somewhat tricky to determine, though perhaps the Unicode confusables database can be a good reference.
> 
> Regarding homoglyphs: I think it’s not just identifiers that can be affected. For example in the accessLevel example in the OP, replacing the “user” string content with some homoglyphs would be equally problematic. There might be more creative uses of this too.

I think we cannot know which characters should be allowed in different contexts (identifiers, string literals, comments...) for every codebase, so something like https://github.com/eslint/eslint/issues/10747 might be the best approach.

**Comment by mhofman** [comment]
> I think we already have a rule that can address this issue: [id-match](https://eslint.org/docs/rules/id-match).

> so something like #10747 might be the best approach.

That rule or any similar that wholesale restrict certain character sets are overly broad and would in effect prevent identifiers which contain characters from certain sets like Cyrillic. Sure a project should be able to decide to enforce such a restriction, but that falls more in the land of style preference.

It wouldn't be a good candidate for shared configurations that should apply broadly to multiple projects, where the language of the author cannot be presumed.


> I think we cannot know which characters should be allowed in different contexts (identifiers, string literals, comments...)

Entirely agree. Identifiers can be understood by tooling. For string literals it might be more complicated, and probably too specific to the codebase. For the homoglyph issue (CVE-2021-42694), I still believe a rule similar to `no-shadow` for identifiers is the right first step. It might be extensible to some strings (e.g. warning that a similar string already appears with a homoglyph, but that's very limited).

**Comment by mdjermanovic** [comment]
> I'm not sure the homoglyph or zero-width-joiner vector deserves a new rule. The `no-shadow` rule already exists and has the same spirit. It could be extended to match identifiers normalizing for these unicode characters. Potentially the `no-redeclare` rule too if we want to keep `no-shadow`'s role to nested scopes.

We can't update these rules to normalize identifiers, as that would mean updating our scope analysis to work incorrectly. 
I believe if ESLint were to treat those identifiers as if they were really same identifiers, that would only add to the confusion and could actually be counterproductive for this issue.

>  I still believe a rule similar to no-shadow for identifiers is the right first step.

A similar rule that would explicitly say what exactly seems wrong in the code makes sense. It could be a rule that reports if there are two different identifiers that have the same or very similar appearance.

> > I think we already have a rule that can address this issue: [id-match](https://eslint.org/docs/rules/id-match).
> 
> > so something like #10747 might be the best approach.
> 
> That rule or any similar that wholesale restrict certain character sets are overly broad and would in effect prevent identifiers which contain characters from certain sets like Cyrillic. Sure a project should be able to decide to enforce such a restriction, but that falls more in the land of style preference.
> 
> It wouldn't be a good candidate for shared configurations that should apply broadly to multiple projects, where the language of the author cannot be presumed.

That's a good point. I think that `id-match` can solve the homoglyph issue for a good number of projects, but a rule that specifically targets the issue (e.g., rule that reports similar identifiers with confusable characters) would be a better candidate for shared configurations. However, such rule sounds to me like a better candidate for a security plugin than for a core ESLint rule.

**Comment by mdjermanovic** [comment]
> disallow bidirectional control characters and homoglyphs in source code to mitigate CVE-2021-42574 and CVE-2021-42694

@joyeecheung can you elaborate more on the proposal about homoglyphs? Is the proposal to entirely disallow certain characters, and which characters would that be?

**Comment by joyeecheung** [comment]
@mdjermanovic I didn't have a specific plan about how homoglyphs should be detected when I opened this - it was mostly about starting a discussion. But my two cent is to use something like http://www.unicode.org/Public/security/8.0.0/confusables.txt to disallow confusable characters of, for example, characters that are more commonly used in code (e.g. ASCII ones). Some clever tricks could be tried to avoid false positives, like checking if adjacent characters are of the same category/subcategory/block. Another more general rule could be to allow users to define certain categories/subcategories/blocks that the characters should be restricted to.

**Comment by mhofman** [comment]
I'd like to propose a multi-pronged approach to mitigate these attacks using eslint rules:
- A rule to disallow bidirectional characters anywhere in a file (comment or code). Similar to `no-tabs` but for these Unicode control characters. Addresses CVE-2021-42574
- A rule to disallow "shadowing" of identifiers containing homoglyphs. This would compute the [skeleton](http://www.unicode.org/reports/tr39/#Confusable_Detection) of identifiers in scope, and check that any declaration of an identifier doesn't have the same skeleton as another identifier in scope. Addresses CVE-2021-42694 for identifiers
- A rule to implement [restriction level detection](http://www.unicode.org/reports/tr39/#Restriction_Level_Detection) for strings, identifiers and properties. This could be used to detect strings with mixed-scripts where they're not expected, or allow projects to enforce latin only if they want. Addresses CVE-2021-42694 for strings or as a more stringent rule for identifiers.

Additionally I'm wondering if we should define an [Identifier Profile](http://www.unicode.org/reports/tr39/#General_Security_Profile) for JavaScript. That might be more difficult than the rules above.

**Comment by nzakas** [comment]
@mhofman i like this approach, though the team feels that the core isn’t the correct home for these rules. I’m talking with the eslint-plugin-security folks to see about getting that project restarted, as I think that would be the best home for such rules.

**Comment by nzakas** [comment]
Definitely still want to look into this, just waiting to hear back from the security plugin folks.

**Comment by LinusU** [comment]
@nzakas did you hear back from the security plugin folks?

**Comment by nzakas** [comment]
Yes, sorry, way too much going on. If you wouldn’t mind, you can suggest security related rules here:

https://github.com/nodesecurity/eslint-plugin-security

Still trying to get back around to fixing up that plugin but it’s the right place for rules like these.

**Comment by nzakas** [comment]
We added a rule for detecting bidi characters to eslint-plugin-security:
https://github.com/eslint-community/eslint-plugin-security/blob/main/rules/detect-bidi-characters.js

@mhofman I know it's been a while, but if you could [propose](https://github.com/eslint-community/eslint-plugin-security/issues/new?assignees=&labels=rule%2Cfeature&template=new-rule.yml&title=New+Rule%3A+%28fill+in%29) the remaining rules to eslint-plugin-security, we'd appreciate it.
